src/editables/__init__.py

```diff
@@ -1,28 +1,37 @@
+import os
+from collections.abc import Iterator
 from pathlib import Path
 
 __all__ = (
     "EditableProject",
     "__version__",
 )
 
-__version__ = "0.2"
+__version__: str = "0.2"
 
 
 class EditableException(Exception):
     pass
 
 
 class EditableProject:
-    def __init__(self, project_name, project_dir):
+    project_name: str
+    project_dir: Path
+    redirections: "dict[str, str]"
+    path_entries: "list[Path]"
+
+    def __init__(
+        self, project_name: str, project_dir: "str | os.PathLike[str]"
```

Comment:
That's a really ugly type, `str | os.PathLike[str]`. Is that really the best way of typing this? And why quote it?

Reply:
This basically says `project_dir` can be a string or something that quacks like a `Path`, i.e. that we can cast to `Path`. It's quoted for the same reason as https://github.com/pfmoore/editables/pull/11#discussion_r659923489.

Reply:
I feel there should be a better way to express "anything that can be cast to a Path". If not, I'm tempted to leave the annotation off and let the later `Path()` call get flagged if what gets passed isn't suitable for passing to `Path`. Maybe that means I'm only interested in type annotations for checking the library code itself, and not for presenting a typed API to callers - that's not a distinction I'd ever really thought about before (none of the typed languages I've used has such a distinction).

I'll need to think a bit more about this...

Reply:
`PathLike` is a protocol with one method, `__fspath__`. Any object which has `__fspath__` can be cast to `Path`.  I'm not sure how else to express this.  We could just have `Union[str, Path]` but that feels "impure", in the same way that you would not annotate a function to take `list`, but `Sequence`.

Reply:
I suspect by defining a named type with a suitable comment

```python
# Why the heck doesn't the stdlib define this FFS :-(
PathLike = Union[str, os.PathLike[str]]
```

Actually, why is it `os.PathLike[str]`? There's no indication in [the documentation](https://docs.python.org/3/library/os.html#os.PathLike) that it should take a type argument, and what would be plausible in there except for str? OK, I'm now going down a rabbit hole, it looks like there's a bunch of issues around this and subscripting `os.PathLike` isn't compatible with Python 3.8:

```
❯ py -3.8
Python 3.8.9 (tags/v3.8.9:a743f81, Apr  2 2021, 11:10:41) [MSC v.1928 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> def f(s: os.PathLike[str]): pass
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'ABCMeta' object is not subscriptable
```

This is more complexity than I'm willing to put up with, and I remain -1 on quoting the annotations (it just looks ugly to me).

I don't want to make a snap decision on this, so I'm going to leave it for a while longer, and likely play around with it myself before deciding. My main priority is not to accept something that I don't understand well enough to support.

Thanks for the PR and opening up the question. I'll get back to you in due course.

Reply:
> I suspect by defining a named type with a suitable comment

I've already done that in the commit to remove the quotes.

> Actually, why is it `os.PathLike[str]`? There's no indication in [the documentation](https://docs.python.org/3/library/os.html#os.PathLike) that it should take a type argument, and what would be plausible in there except for str? OK, I'm now going down a rabbit hole, it looks like there's a bunch of issues around this and subscripting `os.PathLike` isn't compatible with Python 3.8:

Yes, `os.PathLike` is generic in the typeshed but not in the stdlib until 3.9, which is why it needs to be quoted regardless.

Reply:
Looks like this is a well-known issue, [this StackOverflow question](https://stackoverflow.com/questions/55076778/why-isnt-this-function-type-annotated-correctly-error-missing-type-parameters) covers it quite well.

I'm not particularly happy with any of the suggested answers, so until we drop support for older versions of Python (by which I mean "ones that need any of the suggested solutions" 🙂) I'm inclined to leave path arguments untyped (because constraining arguments unnecessarily tightly is IMO unacceptable).

But as I say, I'll let this simmer for a while before decising.

Reply:
In that case it would be better to annotate it as taking simply "object" or "Any". Type checkers expect exported symbols in a `py.typed` package to be annotated in full; if a parameter is not typed, that could be reported as a type error for users of this library.

Reply:
Perhaps annotating 3.6-compatible code is simply not worth the effort.

Reply:
Possibly, although I was surprised that this broke even in 3.8 🙁

I took a look at build, but because it's 2.7 compatible, it's using type comments, so being valid Python syntax isn't required there. The same is true of installer, but it does have a `FSPath` type. But as far as I can see over-constrains (see [here](https://github.com/pradyunsg/installer/blob/main/src/installer/_compat/typing.py#L30)).

Reply:
The typeshed has a large number of internal types which should arguably be made available to all type users but aren't; among them is the `str | os.PathLike[str]` type alias.  But yes, quotes around language constructs are... unsightly, although I don't think that omitting the annotation would be doing anyone a favour.